TE StitchNodes bindings

sjavis opened the pull request:
Added bindings for TempestExtremes StitchNodes. This will be merged after #10 is merged. Closes #15

To do:
- [ ] Rest of the StitchNodes options
- [ ] Error handling
- [ ] Tests
- [ ] Documentation

sjavis commented:
There's a few additional input arguments that aren't defined in the documentation. See if any of these are worth including.

jatkinson1000 commented on docs/developer/index.rst:
````diff
@@ -49,6 +49,7 @@ When reporting a bug, please include the following:
       ```python
       import tctrack
+
       x = 1 + 2
````
My bad, should have caught elsewhere, but since I've seen it here:
```suggestion
      """Your code here."""
```

sjavis commented:
I have finished adding the stitch nodes parameters and documentation. There is currently no testing, but the form of any testing seems like a bit of an open question and also broader than just the stitch nodes bindings.

jatkinson1000 commented on docs/tracking-algorithms/tempest_extremes.rst:
```diff
@@ -59,7 +59,7 @@ Usage of Tempest Extremes in TCTrack is done through the ``tempest_extremes`` mo
 to the methods.
 The detection and stitching algorithm can be configured through the various parameters
-in the 
+in the ``DetectNodesParameters`` and ``StitchNodesParameters`` dataclasses.
```
Whoops, good catch, thanks!

jatkinson1000 commented on docs/tracking-algorithms/tempest_extremes.rst:
```diff
@@ -81,6 +81,11 @@ of one another:
         te.TEContour(var="zgdiff", delta=-6.0, dist=6.5, minmaxdist=1.0),
     ]
 
+    output_commands = [
+        te.TEOutputCommand(var="psl", operator="min", dist=0.0),
+        te.TEOutputCommand(var="orog", operator="max", dist=0.0),
+    ]
+
```
Add something to lines 64-67 above to explain the the example now also adds some output fields (value of psl and orog at the candidate point).

jatkinson1000 commented on src/tctrack/tempest_extremes/tempest_extremes.py:
```diff
@@ -133,6 +133,49 @@ class TEOutputCommand(TypedDict):
+    """
+
+    var: str
+    """Name of the variable being tested. "col" in TempestExtremes."""
```
Maybe clarify this maps to the col option in Tempest Extremes. I initially read it as "col" being the name of the variable that will be tested.

jatkinson1000 commented on src/tctrack/tempest_extremes/tempest_extremes.py:
```diff
@@ -285,27 +328,203 @@ def __str__(self) -> str:
+    in_file : str | None, optional
+        Filename of the DetectNodes output file. If this and `in_list` are ``None``, it
+        will be taken from the DetectNodes parameters. "in" in TempestExtremes.
+    in_list : str | None, optional
+        File containing a list of input files to be processed together.
```
I didn't add this to DetectNodes at present as I think it needs extra documentation specifying how to use it (see our previous discussion on the file needing to be written in 'Tempest Extremes' format.) We could potentially set up something using the `lod_to_te()` function to make this better match the current 'Pythonic' approach we have for `in_file`.

I don't mind leaving this in, but wonder if we should make it unused/unadvised for now.
We should definitely open an issue to add a nice way to use '`in_list`' as a future feature.

jatkinson1000 reviewed (changes requested):
Thanks @sjavis this looks like good work.
I like the reduction into a `_run_te_process()` call.
A few comments/suggestions/discussion points below.

General thoughts:

- `result = TETracker.detect_nodes()` may be misleading, as `result` does not return nodes or tracks but exit code and stderr/out. I wonder if we rename this in the example to something like "`struggling to come up with a good name...`" or make it optional to return results?
- Perhaps the example in the TE docs should be re-worked to show that you do not have to create 2 instances of `TETracker`, one for DN, one for SN, but should instead run both calls of a single object?

sjavis commented on src/tctrack/tempest_extremes/tempest_extremes.py:
```diff
@@ -285,27 +328,203 @@ def __str__(self) -> str:
+        If ``False``, an error is thrown if there are multiple sections in the input
+        nodefile with the same time.
+    caltype : str, default="standard"
+        Type of calendar to use. Options are: ``"standard"``, ``"noleap"``,
```
Yes, it is the Gregorian calendar. I have clarified this.

sjavis commented on src/tctrack/tempest_extremes/tempest_extremes.py:
```diff
@@ -285,27 +328,203 @@ def __str__(self) -> str:
+    add_velocity : bool, default=False
+        Whether to include the velocity components (m/s) of the movement of the TC to
+        the output file.
+    out_file_format : str, default="gfdl"
+        Format of the output file. ``"gfdl"``, ``"csv"``, or ``"csvnoheader"``.
```
Looking into this further the format seems similar (but different) to the output of [TSTORMS](https://github.com/NOAA-GFDL/TCtracker/tree/main/tstorms_driver). So as far as I can tell there is no reference for this format. I will add some information about it to the documentation.

sjavis commented on src/tctrack/tempest_extremes/tempest_extremes.py:
```diff
@@ -285,27 +328,203 @@ def __str__(self) -> str:
+    threshold_filters: list[TEThreshold] | None, optional
+        Filters for paths based on the number of nodes that satisfy a threshold.
+        "thresholdcmd" in TempestExtremes.
+    prioritize : str | None, optional
```
Yes, I was trying to match TE variable names where possible. We can possibly use UK English elsewhere. We should discuss this.

sjavis commented:
I have addressed each of your comments that you made with a number of changes. The main points being:
- Addition to docs showing the preferred use of detect_nodes and stitch_nodes with a single TETracker.
- Changed the variable name for the return of detect/stitch_nodes to run_info to distinguish from the actual output results.
- Removed the tctrack_test.py script
- Added section to the docs describing the default "gfdl" output format given by stitch_nodes.
- Reformatting of docs to fix links and adding some links to parameters/functions/classes

jatkinson1000 commented on docs/api/tempest_extremes_api.rst:
```diff
@@ -8,33 +8,56 @@ It is written by Paul Ullrich and released under a BSD 2-clause license.
    :undoc-members:
    :show-inheritance:
 
+
+.. _stitch-nodes-output-format:
+``stitch_nodes()`` Output Format
+--------------------------------
+
+The default ``"gfdl"`` output of :meth:`TETracker.stitch_nodes` is a plain-text "nodefile" format which
+contains a number of tracks, each of which in the form below.
+
+.. code-block:: text
+
+   start <N> <year> <month> <day> <hour>
+   	 <i> <j> <lon> <lat> <var1> <var2> ... <year> <month> <day> <hour>
+   	 ...
+   	 <i> <j> <lon> <lat> <var1> <var2> ... <year> <month> <day> <hour>
+
+- ``N`` is the number of nodes in the track (and number of lines below header).
+- ``i``, ``j`` are grid indices.
+- ``var1``, ``var2``, etc., are scalar variables as defined by :attr:`StitchNodesParameters.in_fmt` (typically, psl, orog).
+- ``hour`` may instead be seconds if :attr:`StitchNodesParameters.out_seconds` is ``True``.
```
I would also link it from the example in the `Tracking Algorithms` docs.

jatkinson1000 reviewed (commented):
Thanks for addressing those @sjavis 

Just a couple of issues with the rendering of the docs, and a query about where to store the output format information.

sjavis commented on src/tctrack/tempest_extremes/tempest_extremes.py:
```diff
@@ -741,7 +744,7 @@ def detect_nodes(self):
         Examples
         --------
-        To set the parameters, instantiate a ``TETracker`` instance and run
+        To set the parameters, instantiate a :class:``TETracker`` instance and run
```
Fixed
